fix: Add streamIds field to Elasticsearch product mapping

timeo-schmidt opened the pull request:
### 1. Why is this change necessary?

The `streamIds` field is missing from the Elasticsearch product mapping, causing product stream (dynamic product group) filters to fail when Elasticsearch is enabled. This results in 0 products being returned despite products being correctly associated with streams in the database.

Product streams are a core Shopware feature for dynamically grouping products, but when Elasticsearch is enabled, queries filtering by `streamIds` return empty results.

### 2. What does this change do, exactly?

This PR adds the missing `streamIds` field to `ElasticsearchProductDefinition` in three places:

1. **Mapping Definition** (`getMapping()` method): Adds `'streamIds' => self::KEYWORD_FIELD` to the properties array
2. **Document Building** (`fetch()` method): Adds `'streamIds' => ElasticsearchIndexingUtils::parseJson($item, 'streamIds')` to parse and include stream IDs in indexed documents
3. **SQL Query** (`fetchProducts()` method): Adds `IFNULL(p.stream_ids, pp.stream_ids) AS streamIds` to fetch stream IDs from the database

The field follows the same pattern as other ID array fields (`categoryIds`, `propertyIds`, `optionIds`, `tagIds`) and is indexed as a `KEYWORD_FIELD` for exact matching.

**Testing:**
- Extended `testMapping()` to verify `streamIds` is included in the expected mapping
- Extended `testFetching()` to verify `streamIds` is correctly fetched and included in documents
- Added mock data with realistic hex IDs to verify the complete data flow

### 3. Describe each step to reproduce the issue or behaviour.

**Without this fix:**

1. Enable Elasticsearch in Shopware:
   ```ini
   SHOPWARE_ES_ENABLED=1
   SHOPWARE_ES_INDEXING_ENABLED=1
   ```

2. Create a product stream with products assigned to it

3. Index products: `php bin/console es:index`

4. Query product listing API with a streamIds filter:
   ```bash
   curl 'https://example.com/store-api/product-listing/{categoryId}' \
     -H 'Content-Type: application/json' \
     -d '{"filter": [{"type": "contains", "field": "streamIds", "value": "{streamId}"}]}'
   ```

5. **Result:** `{"elements": [], "total": 0}` ❌ (0 products returned)

6. Disable Elasticsearch (`SHOPWARE_ES_ENABLED=0`) and repeat query → Returns correct number of products ✅

**With this fix:**

After applying the fix and reindexing:
```bash
php bin/console es:mapping:update
php bin/console es:index
```

The same query now returns the correct number of products.

**Note:** Shops using Elasticsearch need to update the mapping and reindex after upgrading. No database migrations required.

### 4. Please link to the relevant issues (if any).

Fixes: #13151

<!-- Replace XXXXX with the actual GitHub issue number -->

### 5. Checklist

- [x] I have written tests and verified that they fail without my change
- [x] I have created a [changelog file](https://github.com/shopware/shopware/blob/trunk/adr/2020-08-03-implement-new-changelog.md) with all necessary information about my changes
- [x] I have written or adjusted the documentation according to my changes (not required)
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfilled them

### Backport
A backport of this would be desirable, as it affects 6.6.X / 6.5.X in the same way.

timeo-schmidt committed:
fix: Add streamIds field to Elasticsearch product mapping

- Added missing streamIds field to ElasticsearchProductDefinition mapping
- Field is now indexed as KEYWORD_FIELD like other ID arrays
- Added streamIds to document building in fetch() method
- Added streamIds to SQL query in fetchProducts() method
- Added test coverage for streamIds in mapping and fetch operations

This fix enables product stream (dynamic product group) filtering
when Elasticsearch is enabled, resolving 0-result queries when
filtering by streamIds.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/shopware/shopware?pullRequest=13153) <br/>All committers have signed the CLA.

timeo-schmidt committed:
Merge branch 'trunk' into fix/elasticsearch-streamids-mapping

vienthuong commented on changelog/_unreleased/2025-10-17-add-streamids-to-elasticsearch-product-mapping.md:
```diff
@@ -0,0 +1,54 @@
+
+# Core
+
+-   Added missing `streamIds` field to Elasticsearch product mapping in `ElasticsearchProductDefinition`
```
Could you reformat this changelog file a bit? 

- reindent lines
- only keywords Added/Changed/Removed/Deprecated are allowed 
- omit no changes sections

timeo-schmidt committed:
docs: Reformat changelog according to maintainer feedback

- Reindent lines with * instead of -
- Use only allowed keywords (Added)
- Remove "no changes" sections
- Consolidate related changes

vienthuong reviewed (approved):
beside some very subtle remark, the change LGTM, thank you for your contribution 🎉

vienthuong commented on changelog/_unreleased/2025-10-17-add-streamids-to-elasticsearch-product-mapping.md:
```diff
@@ -0,0 +1,28 @@
+# Core
+* Added missing `streamIds` field to Elasticsearch product mapping in `ElasticsearchProductDefinition`
+* Added `KEYWORD_FIELD` indexing for `streamIds` consistent with other ID array fields (`categoryIds`, `propertyIds`, `optionIds`, `tagIds`)
+
```
```
___
```

vienthuong commented on changelog/_unreleased/2025-10-17-add-streamids-to-elasticsearch-product-mapping.md:
```diff
@@ -0,0 +1,28 @@
+
+# API
+* Added support for filtering product listings by `streamIds` when using Elasticsearch
+
```
```
___
```

timeo-schmidt committed:
docs: Add section dividers to changelog

mitelg commented on changelog/_unreleased/2025-10-17-add-streamids-to-elasticsearch-product-mapping.md:
```diff
@@ -0,0 +1,32 @@
+---
+title: Add streamIds field to Elasticsearch product mapping
+issue: GITHUB-13151
+flag: skip-trigger-flow
```
```suggestion
```

not a valid issue, not a valid flag. please review your own changes and do not let AI do all the stuff without re-checking

mitelg commented on changelog/_unreleased/2025-10-17-add-streamids-to-elasticsearch-product-mapping.md:
```diff
@@ -0,0 +1,32 @@
+
+# Core
+* Added missing `streamIds` field to Elasticsearch product mapping in `ElasticsearchProductDefinition`
+* Added `KEYWORD_FIELD` indexing for `streamIds` consistent with other ID array fields (`categoryIds`, `propertyIds`, `optionIds`, `tagIds`)
```
```suggestion
* Added missing `streamIds` field to Elasticsearch product mapping in `ElasticsearchProductDefinition`
```

mitelg commented on changelog/_unreleased/2025-10-17-add-streamids-to-elasticsearch-product-mapping.md:
```diff
@@ -0,0 +1,32 @@
+
+---
+
+# Upgrade Information
```
```suggestion

___

# Upgrade Information
```

mitelg commented on changelog/_unreleased/2025-10-17-add-streamids-to-elasticsearch-product-mapping.md:
````diff
@@ -0,0 +1,32 @@
+php bin/console es:index
+```
+
+This fix resolves an issue where product stream (dynamic product group) filters would return 0 results when Elasticsearch was enabled, even though products were correctly associated with streams in the database.
````
```suggestion
## ProductStream IDs added to ElasticsearchProductDefinition

Product streams are now supported while using Elasticsearch. To make this work, a re-index is necessary. 
```

mitelg committed:
Merge branch 'trunk' into fix/elasticsearch-streamids-mapping

timeo-schmidt committed:
docs: apply suggestions from code review

Co-authored-by: Michael Telgmann <mitelg@users.noreply.github.com>

timeo-schmidt committed:
docs: fixing github issue syntax and removing unnecessary flag

vienthuong committed:
Merge branch 'trunk' into fix/elasticsearch-streamids-mapping

vienthuong commented on changelog/_unreleased/2025-10-17-add-streamids-to-elasticsearch-product-mapping.md:
```diff
@@ -0,0 +1,18 @@
+---
+title: Add streamIds field to Elasticsearch product mapping
```
Hi @timeo-schmidt, sorry for the delayed response, with our new releases process, this changelog is no longer necessary, if it's needed for devs or worth noting in the release you can add this point `ProductStream IDs added to ElasticsearchProductDefinition` to `[RELEASE_INFO-6.7.md](https://github.com/shopware/shopware/blob/trunk/RELEASE_INFO-6.7.md)`

vienthuong committed:
Merge branch 'trunk' into fix/elasticsearch-streamids-mapping

nguyenquocdaile committed:
chore: update  release info #13153

mitelg commented on RELEASE_INFO-6.7.md:
```diff
@@ -294,3 +294,4 @@ https://github.com/shopware/shopware/pull/12654 by ffrank913 fixes Incorrect fro
   * Removal of `controllerName` and `controllerAction` variables in templates
   * Deprecation of `Shopware\Core\System\SalesChannel\SalesChannel\SalesChannelContextSwitcher`
 * Upgrade notes: DB migration for the new category index, admin build target upgrade, profiler defaults
+* ProductStream IDs added to ElasticsearchProductDefinition
```
this were the examples for 6.7.3.0

as we are preparing 6.7.5.0 already, this needs to be put at the top of this file

nguyenquocdaile committed:
Merge branch 'trunk' into fix/elasticsearch-streamids-mapping

nguyenquocdaile commented:
Hi @timeo-schmidt , Your PR is merged. Thanks for your contribution! 💙
